File: scripts/form_issue.py
import re
import json
import uuid
import sys
from pathlib import Path
from jsonschema import validate, ValidationError

# Read issue body (markdown) from file
issue_file = Path(sys.argv[1])
if not issue_file.exists():
    print(f"❌ File not found: {issue_file}")
    sys.exit(1)

body = issue_file.read_text()

def extract(label):
    pattern = rf"### {re.escape(label)}\s+(.+?)(?=\n###|\Z)"
    match = re.search(pattern, body, re.DOTALL)
    return match.group(1).strip() if match else ""

# Extract event fields
event = {
    "title": extract("Event Title"),
    "date": extract("Event Date (ISO format)"),
    "startTime": extract("Start Time (UTC)"),
    "endTime": extract("End Time (UTC)"),
    "country": extract("Country"),
    "city": extract("City"),
    "eventType": extract("Event Type"),
    "gatheringType": extract("Gathering Type"),
    "ticket": extract("Ticket Required?"),
    "capacity": extract("Capacity"),
    "description": extract("Description"),
    "link": extract("Event Link"),
    "organizer": {
        "name": extract("Organizer Name"),
        "twitter": extract("Organizer Twitter Handle (optional)").replace("_No response_", "").strip()
    },
    "id": str(uuid.uuid4())
}

# Validate against schema
with open("schema.json") as f:
    schema = json.load(f)

try:
    validate(instance=event, schema=schema)
except ValidationError as e:
    print("❌ Schema validation error:")
    print(e)
    sys.exit(1)

# Load and update events.json
events_file = Path("events.json")
events = []
if events_file.exists():
    events = json.loads(events_file.read_text())
events.append(event)

with open(events_file, "w") as f:
    json.dump(events, f, indent=2)

print("✅ Event added and validated")

# Generate Markdown PR body
pr_body = f"""## 🆕 New WWDC Community Event Submitted

| Field              | Value |
|-------------------|-------|
| **Title**          | {event['title']} |
| **Date**           | {event['date']} |
| **Start Time**     | {event['startTime']} |
| **End Time**       | {event['endTime']} |
| **Country**        | {event['country']} |
| **City**           | {event['city']} |
| **Event Type**     | {event['eventType']} |
| **Gathering Type** | {event['gatheringType']} |
| **Ticket**         | {event['ticket']} |
| **Capacity**       | {event['capacity']} |
| **Description**    | {event['description']} |
| **Link**           | [{event['link']}]({event['link']}) |
| **Organizer**      | {event['organizer']['name']} |
| **Twitter**        | {event['organizer']['twitter'] or '*(none)*'} |

---

✅ This event has been auto-validated and added to `events.json`.  
Maintainers can review and merge the PR to publish it in the live feed.
"""

# Write PR body to tmp/pr-body.txt
tmp_dir = Path("tmp")
tmp_dir.mkdir(exist_ok=True)
with open(tmp_dir / "pr-body.txt", "w") as f:
    f.write(pr_body)

print("📄 PR body saved to tmp/pr-body.txt")

File: scripts/form-issue.js
const fs = require('fs');
const crypto = require('crypto');

// Get the issue JSON file path from command line arguments
const issueFilePath = process.argv[2];

if (!issueFilePath) {
  console.error('Usage: node from-issue.js <issue-json-file>');
  process.exit(1);
}

// Read the issue body from the JSON file
let issueBody;
try {
  issueBody = JSON.parse(fs.readFileSync(issueFilePath, 'utf8'));
} catch (error) {
  console.error('Error reading issue file:', error);
  process.exit(1);
}

// Function to extract field values from the issue body
function extractField(body, fieldName) {
  // Match patterns like "### Event Title\n\nVision Pro Developer Meetup"
  const regex = new RegExp(`### ${fieldName}\\s*\\n\\s*([^\\n#]+)`, 'i');
  const match = body.match(regex);
  return match ? match[1].trim() : null;
}

// Extract all the fields
const event = {
  title: extractField(issueBody, "Event Title"),
  date: extractField(issueBody, "Event Date \\(ISO format\\)"),
  startTime: extractField(issueBody, "Start Time \\(UTC\\)"),
  endTime: extractField(issueBody, "End Time \\(UTC\\)"),
  country: extractField(issueBody, "Country"),
  city: extractField(issueBody, "City"),
  eventType: extractField(issueBody, "Event Type"),
  gatheringType: extractField(issueBody, "Gathering Type"),
  ticket: extractField(issueBody, "Ticket Required\\?"),
  capacity: extractField(issueBody, "Capacity"),
  description: extractField(issueBody, "Description"),
  link: extractField(issueBody, "Event Link"),
  organizer: {
    name: extractField(issueBody, "Organizer Name"),
    twitter: extractField(issueBody, "Organizer Twitter Handle \\(optional\\)")
  },
  id: crypto.randomUUID()
};

// Create PR body with the event data in the format expected by process-form.js
let prBody = `Adding new WWDC community event

**Event Title:**
<!--${event.title}-->

**Date (ISO format):**
<!--${event.date}-->

**Start Time (UTC):**
<!--${event.startTime}-->

**End Time (UTC):**
<!--${event.endTime}-->

**Country:**
<!--${event.country}-->

**City:**
<!--${event.city}-->

**Event Type:**
<!--${event.eventType}-->

**Gathering Type:**
<!--${event.gatheringType}-->

**Ticket Required?:**
<!--${event.ticket}-->

**Capacity:**
<!--${event.capacity}-->

**Event Description:**
<!--${event.description}-->

**Link:**
<!--${event.link}-->

**Organizer Name:**
<!--${event.organizer.name}-->

**Twitter (optional):**
<!--${event.organizer.twitter || ''}-->
`;

// Create or update events.json
let events = [];
try {
  if (fs.existsSync('events.json')) {
    events = JSON.parse(fs.readFileSync('events.json', 'utf8'));
  }
} catch (error) {
  console.log('No existing events.json found or error reading it. Creating new one.');
}

// Add new event
events.push(event);

// Save updated events to events.json
fs.writeFileSync('events.json', JSON.stringify(events, null, 2));

// Also save the PR body to a file for the GitHub action to use
fs.writeFileSync('tmp/pr-body.txt', prBody);

console.log('✅ Event processed and added to events.json');
console.log('✅ PR body created');